### PR TITLE
format: add checks for type alias

### LINKED ISSUE #9840: format: add checks for type aliases
*Description*:
Envoy has several conventions around type aliases defined in STYLE.md. Right now these conventions are checked by reviewers. It would be really useful to have the conventions checked in the format CI run.

This could be done with regexes in the check_format.py script, or optimally by extending clang-tidy.

### PR DESCRIPTION
**Pull request by tomocy**
Signed-off-by: tomocy <tomocy.dev@gmail.com>

Commit Message: format: add checks for type alias
Additional Description: N/A
Risk Level: Low
Testing: N/A
Docs Changes: N/A
Release Notes: N/A
Fixes #9840

### CODE REVIEW

File: tools/code_format/check_format.py [hunk 1]
```diff
@@ -696,6 +696,15 @@ def checkSourceLine(line, file_path, reportError):
         reportError("Don't call grpc_init() or grpc_shutdown() directly, instantiate " +
                     "Grpc::GoogleGrpcContext. See #8282")
 
+  if not line.startswith("using"):
+    smart_ptr_m = re.search("std::(unique_ptr|shared_ptr)<(.*?)>", line)
```
Comment (htuch) [inline tools/code_format/check_format.py:700]: Is this a bit too strict? I understand now wanting to have verbose unique_ptr for commonly used types, but the occasional use of unique_ptr, in particular for non-Envoy types, makes sense.

Comment (tomocy) [reply tools/code_format/check_format.py:700]: I’m going to add an allowed type list (eg. [int, std::string]) and report an error for the types which are not included in that list. I suppose it handles normal use cases well.

File: tools/code_format/check_format.py [hunk 2]
```diff
@@ -352,6 +398,9 @@ def whitelistedForUnpackTo(file_path):
       "./source/common/protobuf/utility.cc", "./source/common/protobuf/utility.h"
   ]
 
+def whitelistedForNonTypeAlias(name):
+  return re.match(r".*\[\]$", name) or re.match("^std::vector<.*", name) or re.match(f"^({'|'.join(NON_TYPE_ALIAS_ALLOWED_TYPES)})$", name)
```
[the remark this answers is not in the document]
Comment (tomocy) [reply tools/code_format/check_format.py]: done

File: tools/testdata/check_format/using_type_alias.cc [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+class Connection;
+
+using ConnectionPtr = std::unique_ptr<Connection>;
+
+class A {
+  using ConnectionSharedPtr = std::shared_ptr<Connection>;
+  using ConnectionOptRef = absl::optional<std::reference_wrapper<Connection>>;
+};
+} // namespace Network
+} // namespace Envoy
```
[the remark this answers is not in the document]
Comment (tomocy) [reply tools/testdata/check_format/using_type_alias.cc]: done

**Comment by tomocy** [comment]
Thank you for your review. I considered adding `--aggressive` flag but this flag would not be enabled as long as those types reported in CI are aliased. I'm looking for a tool or a way that automatically add type alias and is invoked by `check_format.py fix`. Thinking that the code format fixer never breaks code, It is also good to run the tool before this change is merged.

**Comment by jmarantz** [comment]
I'm having trouble parsing your phrase:
> but this flag would not be enabled as long as those types reported in CI are aliased

That makes me want to clarify a few things.
1. This PR cannot be merged until all the format errors are resolved. You'll want automation to do that. There are lots of examples in check_format.py where fixes are applied automatically.
2. The `--aggressive` flag would be set by a human when running `check_format fix --aggressive`, and the idea here would be that the PR might be left in an unbuildable state, which you would then manually fix and test before committing.

Does that makes sense?

File: tools/code_format/check_format.py [hunk 4]
```diff
@@ -696,6 +747,17 @@ def checkSourceLine(line, file_path, reportError):
         reportError("Don't call grpc_init() or grpc_shutdown() directly, instantiate " +
                     "Grpc::GoogleGrpcContext. See #8282")
 
+  if not re.search("using .* = .*;", line):
```
Comment (jmarantz) [inline tools/code_format/check_format.py:750]: precompile all the regexes.

File: tools/code_format/check_format.py [hunk 5]
```diff
@@ -696,6 +747,17 @@ def checkSourceLine(line, file_path, reportError):
...
+    smart_ptr_m = re.finditer("std::(unique_ptr|shared_ptr)<(.*?)>", line)
```
Comment (jmarantz) [inline tools/code_format/check_format.py:751]: what does the `_m` mean? Spell that out?

**Comment by tomocy** [comment]
I mean 1. I think that it needs C++ AST to resolve those errors perfectly and is difficult to do only with Python scripts. I am looking for a way to use the AST and fix it, and if it is not possible, I will replace those lines with scripts and declare type aliases manually.

File: tools/code_format/check_format.py (hunk 5 above)
Comment (tomocy) [reply tools/code_format/check_format.py]: done

File: tools/code_format/check_format.py (hunk 4 above)
Comment (tomocy) [reply tools/code_format/check_format.py]: done

**Comment by tomocy** [comment]
I have changed my mind and let it check the format in the added line as pre-push.

File: support/hooks/pre-push [hunk 6]
```diff
@@ -14,7 +14,7 @@ DUMMY_SHA=0000000000000000000000000000000000000000
 echo "Running pre-push check; to skip this step use 'push --no-verify'"
 
 while read LOCAL_REF LOCAL_SHA REMOTE_REF REMOTE_SHA
-do
+do
```
Comment (jmarantz) [inline support/hooks/pre-push:17]: remove trailing whitespace

File: tools/code_format/check_format.py [hunk 7]
```diff
@@ -168,6 +168,7 @@
   "extensions/filters/network/common",
   "extensions/filters/network/common/redis",
 }
+
```
Comment (jmarantz) [inline tools/code_format/check_format.py:171]: revert

**Comment by tomocy** [comment]
Thank you for your advice. I am adding `—aggressive` flag.

**Comment by jmarantz** [comment]
merge master needed. ping when ready for review.

/wait

**Comment by tomocy** [comment]
@jmarantz ping. It is getting ready to be reviewed.

File: source/common/buffer/buffer_impl.h [hunk 8]
```diff
@@ -484,6 +489,8 @@ class BufferFragmentImpl : NonCopyable, public BufferFragment {
   const std::function<void(const void*, size_t, const BufferFragmentImpl*)> releasor_;
 };
 
+using BufferFragmentImplPtr = std::unique_ptr<BufferFragmentImpl>;
```
Comment (jmarantz) [inline source/common/buffer/buffer_impl.h:492]: it's not obvious to me why this was needed. It does not appear to be used.

File: source/common/buffer/buffer_impl.h [hunk 9]
```diff
@@ -597,6 +604,8 @@ class OwnedImpl : public LibEventInstance {
   OverflowDetectingUInt64 length_;
 };
 
+using OwnedImplPtr = std::unique_ptr<OwnedImpl>;
```
Comment (jmarantz) [inline source/common/buffer/buffer_impl.h:607]: is this used?

**Comment by tomocy** [comment]
I added type aliases every the compiler failed to build due to the undefined error. But I did it manually so a part of those may not be used.

My understanding is that I will checkout two kinds of branches where I cherry-pick the changes about type aliasing by 50 file fragments while I push the changes about Python tools.

Is it correct?

**Comment by jmarantz** [comment]
RE "unused type declarations": I thought we had talked about an automated approach to resolving these (even that automation was not robust enough to never fail). Sounds like you did a lot of manual work. Sorry about that. If we wind up with extra type aliases I won't get too worried.

RE "branches and cherry-picks": I am really barely functional at git and do things in kind of a low-tech way. But what I would do is something like:

1. merge master
2. git diff >/tmp/all_changes.diff
3. edit all_changes.diff in a text editor, splitting it into K digestible chunks, especially the one with the tooling chnages
4. open up K new branches from master, and patch each of the diff files, e.g. with "patch -p1 </tmp/changes_23.diff"
5. make sure each of the branches pass tests on their own with "bazel test //test/...". 
6. start streaming out reviews.

Even if you wanted to make K be fairly small that would be fine; the one I want to review first is the one with the tooling changes, even though we won't be able to merge it until all the others are done.

Does that make sense?

**Comment by tomocy** [comment]
Thank you, it makes sense. I am trying in the way.

**Pull request closed without merging**